### PR TITLE
[Kernel Selection] Remove hard code of PADDLE_WITH_MKLDNN

### PR DESCRIPTION
**Pull request by jiahy0825**
### PR types
Function optimization

### PR changes
OPs

#### 1. Background
Op relies on the function `GetExpectedKernelType` to get its target kernel type `OpKernelType` and then matches suitable kernel. When MKLDNN kernel is registered in Paddle, we have to add hard code `#ifdef PADDLE_WITH_MKLDNN` into `GetExpectedKernelType` to indicate MKLDNN kernel explicitly. However, such implementation isn't elegant, which adds many repetitious codes into the framework (only a few MKLDNN kernels need specific logic) and makes Paddle hard to maintain. Thus we decide to optimize the implementation of `GetExpectedKernelType`.

#### 2. Goal
Remove the hard code of `#ifdef PADDLE_WITH_MKLDNN` within `GetExpectedKernelType`, keep the implementation of `GetExpectedKernelType` simple and clean. Such change will help a lot when we support the MKLDNN kernel dispatch process of `PHI` in the near future.

#### 3. Main Change
1. Delete the repetitious codes of `#ifdef PADDLE_WITH_MKLDNN` and move such logic outside `GetExpectedKernelType` (this PR).
2. Handle ops which have both`#if defined(PADDLE_WITH_CUDA) || defined(PADDLE_WITH_HIP)` and `#ifdef PADDLE_WITH_MKLDNN` in `GetExpectedKernelType` (this PR). 
3. **Next step**: deal with ops that have specific implementations and ops with customized type values.

#### 4. Relevant Ops of This PR: 
`abs`, `abs_grad`, `batch_norm`, `batch_norm_grad`, `clip`, `clip_grad`, `concat`, `concat_grad`, `data_norm`, `data_norm_grad`, `expand_v2`, `expand_v2_grad`, `fill_constant`, `fusion_gru`, `fusion_lstm`, `gaussian_random`, `log_softmax`, `lrn`, `lrn_grad`, `matmul`, `matmul_grad`, `matmul_v2`, `matmul_v2_grad`, `scale`, `shape`, `shuffle_channel`, `stack`, `transpose`, `transpose_grad`, `transpose2_grad`, `gelu`, `gelu_grad`, `prelu`, `prelu_grad`, `nearest_interp`, `nearest_interp_v2`, `bilinear_interp`, `bilinear_interp_v2`, `quantize`, `requantize`, `dequantize`, `conv2d_transpose`, `softmax`, `softmax_grad`
`elementwise relevant ops (ElementwiseOp, ElementwiseOpGrad, etc.)`
`activation relevant ops (ActivationOp, ActivationOpGrad, etc.)`

#### 1. 背景
Op 依赖函数 `GetExpectedKernelType` 来获取其对应的 kernel 类型 `OpKernelType`，然后再匹配到合适的 kernel。 MKLDNN 的 kernel 在 Paddle 中注册时，我们必须在 `GetExpectedKernelType` 中添加硬编码 `#ifdef PADDLE_WITH_MKLDNN` 以明确指示 MKLDNN 内核。 然而，这样的实现不够优雅，它在框架中添加了许多重复的代码（只有少数 MKLDNN 内核需要特化的处理逻辑）并且使得 Paddle 难以维护。 因此我们决定优化 `GetExpectedKernelType` 的实现。

#### 2. 目标
移除 `GetExpectedKernelType` 中的 `#ifdef PADDLE_WITH_MKLDNN` 硬编码，保持一个简单干净的 `GetExpectedKernelType` 函数实现。 本 PR 也有助于支持后续 phi 中 mkldnn kernel 的选择分发机制。

#### 3. 主要改动
1. 删除 `#ifdef PADDLE_WITH_MKLDNN` 的重复代码，并将该逻辑移到 `GetExpectedKernelType` 之外（本PR）
2. 处理同时具有`#if defined(PADDLE_WITH_CUDA) || defined(PADDLE_WITH_HIP)` 和 `#ifdef PADDLE_WITH_MKLDNN` 硬编码的op（本PR）
3. **下一步规划**：处理具有特化逻辑和自定义类型值的op。

### CODE REVIEW

**Review by veyron95 — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/operator.cc [hunk 1]
```diff
@@ -1776,6 +1816,23 @@ void OperatorWithKernel::RunImpl(const Scope& scope,
...
+// GetExpectedKernelType, so that if MKLDNN can be used, the library_type_ and
+// data_layout_ of expected_kernel_key need to be adjusted. There are three
+// statements in if condition: The first statement checks whether library_type_
+// are changed by other high priority backends; the second checks whether this
+// op has specific implementation; the third checks whether mkldnn kernel can be
+// used.
+#ifdef PADDLE_WITH_MKLDNN
+  if (expected_kernel_key.library_type_ == framework::LibraryType::kPlain &&
+      !paddle::platform::in_mkldnn_white_list(type_) &&
+      this->CanMKLDNNBeUsed(ctx, expected_kernel_key.data_type_)) {
```
Comment (zyfncg) [inline paddle/fluid/framework/operator.cc:1825]: 这里如果是每个算子执行时都会进行一遍判断的话可能会对性能有一定影响，也许后面需要关注下

**Review by YuanRisheng — DISMISSED** [review]

**Review by zyfncg — DISMISSED** [review]

File: paddle/fluid/framework/operator.cc (hunk 1 above)
Comment (jiahy0825) [reply paddle/fluid/framework/operator.cc:1825]: 是的，感谢～
Notice: these codes may lead to performance loss, which we should take care of in the future.

File: paddle/fluid/imperative/execution_context.h [hunk 2]
```diff
@@ -102,7 +102,10 @@ class DygraphExecutionContext : public framework::ExecutionContext {
   }
 
   bool HasAttr(const std::string& name) const override {
-    return attrs_.count(name) != 0 || default_attrs_.count(name) != 0;
+    if (attrs_.find(name) == attrs_.end()) {
```
Comment (chenwhql) [inline paddle/fluid/imperative/execution_context.h:105]: Is the logic before and after the modification the same?

File: paddle/fluid/platform/mkldnn_op_list.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chenwhql) [inline paddle/fluid/platform/mkldnn_op_list.h:1]: The license requires proper blank line and indentation

**Review by chenwhql — DISMISSED** [review]

File: paddle/fluid/imperative/execution_context.h (hunk 2 above)
Comment (jiahy0825) [reply paddle/fluid/imperative/execution_context.h:105]: Yeah, they are the same logic.
Generally, count() will be implemented using find(), so here I use find() directly.

File: paddle/fluid/platform/mkldnn_op_list.h (hunk 3 above)
Comment (jiahy0825) [reply paddle/fluid/platform/mkldnn_op_list.h:1]: Done, thx~

**Comment by jiahy0825** [comment]
@jczaja @Silv3S could you please review this PR?

**Review by Silv3S — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
